### PR TITLE
Weekly comparison - average text summarization

### PR DESCRIPTION
**Pull request by jcc37**
Add `TrendWeeklyComparisonAverage` summarization module and tests for time-series data.
The summaries describe the comparison between average y-values of every two consecutive weeks. The following are sample summaries for a one-month time-series data:
* The average active users of the second week was 24.3% more than the first week.
* The average active users of the third week was 12.2% more than the second week.
* The average active users of the fourth week was 17.1% more than the third week.

### CODE REVIEW

File: src/datasets/summarizations/libs/trend.ts [hunk 1]
```diff
@@ -141,3 +148,69 @@ export function exponentialMovingAverage(points: TimeSeriesPoint[], alpha = 0.3)
   }));
   return smoothedPoints;
 }
+
+export function centeredMovingAverage(points: TimeSeriesPoint[], k: number): TimeSeriesPoint[] {
```
Comment (Rhombus33) [inline src/datasets/summarizations/libs/trend.ts:152]: Add description and rename function, maybe generateCenteredMovingAveragePoints?

File: src/datasets/summarizations/libs/trend.ts [hunk 2]
```diff
@@ -141,3 +148,69 @@ export function exponentialMovingAverage(points: TimeSeriesPoint[], alpha = 0.3)
   }));
   return smoothedPoints;
 }
+
+export function centeredMovingAverage(points: TimeSeriesPoint[], k: number): TimeSeriesPoint[] {
+  const L = points.length;
```
Comment (Rhombus33) [inline src/datasets/summarizations/libs/trend.ts:153]: Use full words numOfPoints or lengthOfPoints

File: src/datasets/summarizations/libs/trend.ts [hunk 3]
```diff
@@ -141,3 +148,69 @@ export function exponentialMovingAverage(points: TimeSeriesPoint[], alpha = 0.3)
   }));
   return smoothedPoints;
 }
+
+export function centeredMovingAverage(points: TimeSeriesPoint[], k: number): TimeSeriesPoint[] {
+  const L = points.length;
+  const smoothedPoints: TimeSeriesPoint[] = [];
+  for (let i = 0; i < points.length; i++) {
+    const lSumPoints = points.slice(Math.max(0, i - k), Math.min(L, i + k));
```
Comment (Rhombus33) [inline src/datasets/summarizations/libs/trend.ts:156]: what are l and r?

File: src/datasets/summarizations/libs/trend.ts [hunk 4]
```diff
@@ -141,3 +148,69 @@ export function exponentialMovingAverage(points: TimeSeriesPoint[], alpha = 0.3)
...
+    const smoothedY = 0.5 * (lSum / lSumPoints.length + rSum / rSumPoints.length);
+    smoothedPoints.push({
+      x: points[i].x,
+      y: smoothedY,
+    });
+  }
+  return smoothedPoints;
+}
+
+export type GroupIdentifier = string | number;
```
Comment (Rhombus33) [inline src/datasets/summarizations/libs/trend.ts:171]: This name is very generic, could you add a comment. Is this a unique id? Why do we need a unique id?

File: src/datasets/summarizations/libs/trend.ts [hunk 5]
```diff
@@ -141,3 +148,69 @@ export function exponentialMovingAverage(points: TimeSeriesPoint[], alpha = 0.3)
...
+export interface DecompositionResult {
```
Comment (Rhombus33) [inline src/datasets/summarizations/libs/trend.ts:172]: Please add description to interface and properties. Are all the points separated into one of these categories? Can a point be in more than one category? What do these categories mean?

File: src/datasets/summarizations/libs/trend.ts [hunk 6]
```diff
@@ -141,3 +148,69 @@ export function exponentialMovingAverage(points: TimeSeriesPoint[], alpha = 0.3)
...
+  detrendPoints: TimeSeriesPoint[];
+  seasonPoints: TimeSeriesPoint[];
+  residualPoints: TimeSeriesPoint[];
+}
+
+export function additiveDecomposition(
```
Comment (Rhombus33) [inline src/datasets/summarizations/libs/trend.ts:178]: Add description and rename function

File: src/datasets/summarizations/libs/trend.ts [hunk 7]
```diff
@@ -141,3 +148,69 @@ export function exponentialMovingAverage(points: TimeSeriesPoint[], alpha = 0.3)
...
+
+  const detrendPoints = points.map(({ x, y }, i) => ({
+    x,
+    y: y - trendPoints[i].y,
+  }));
+
+  const groups: Record<GroupIdentifier, TimeSeriesPoint[]> = {};
+
+  for (const point of detrendPoints) {
+    const gid = groupFn(point);
```
Comment (Rhombus33) [inline src/datasets/summarizations/libs/trend.ts:191]: What is gid? Is it groupId? If so, use the full name

File: src/datasets/summarizations/libs/trend.ts [hunk 8]
```diff
@@ -141,3 +148,69 @@ export function exponentialMovingAverage(points: TimeSeriesPoint[], alpha = 0.3)
...
+    groups[gid] = [point, ...(groups[gid] ?? [])];
+  }
+
+  const groupAverages: Record<GroupIdentifier, number> = {};
+  for (const [gid, gpoints] of Object.entries(groups)) {
```
Comment (Rhombus33) [inline src/datasets/summarizations/libs/trend.ts:196]: Use full names as variable names

File: src/datasets/summarizations/libs/trend.ts [hunk 9]
```diff
@@ -141,3 +148,69 @@ export function exponentialMovingAverage(points: TimeSeriesPoint[], alpha = 0.3)
...
+    const s = sum(gpoints.map(({ y }) => y));
```
Comment (Rhombus33) [inline src/datasets/summarizations/libs/trend.ts:197]: Use full names for variable names

File: src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
+
+import { Summary } from './types';
+import { TimeSeriesPoint } from '../metas/types';
+import { cacheSummaries } from './utils/commons';
+import {
+  groupPointsByXWeek,
+} from './utils/time-series';
+import { formatY } from '../../utils/formatters';
+
+export function queryFactory(points: TimeSeriesPoint[]) {
```
Comment (Rhombus33) [inline src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts:10]: Add a description overview for this function

File: src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+import { cacheSummaries } from './utils/commons';
+import {
+  groupPointsByXWeek,
+} from './utils/time-series';
+import { formatY } from '../../utils/formatters';
+
+export function queryFactory(points: TimeSeriesPoint[]) {
+  return cacheSummaries(() => {
+    // Only consider weeks with more than 3 days when creating summaries
+    // Weeks with 3 days or less are considered to belong to last/next 30 days
```
Comment (Rhombus33) [inline src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts:13]: Not sure I understand. Are weeks with < 3 days omitted in this summary OR used as a comparison "last/next/30 days"

File: src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    const weekPointArrays = groupPointsByXWeek(points).filter(weekPoints => weekPoints.length >= 4);
+    const nWeeks = weekPointArrays.length;
```
Comment (Rhombus33) [inline src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts:15]: numOfWeeks

File: src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+    const weekAverages = weekPointArrays.map(weekPoints => weekPoints.reduce((p, { y }) => p + y, 0) / weekPoints.length);
```
Comment (Rhombus33) [inline src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts:17]: "weeklyAverages" I think you are referring to multiple weeks here.

File: src/utils/misc.ts [hunk 14]
```diff
@@ -20,6 +20,15 @@ export function mod(a: number, n: number) {
   return (a % n + n) % n;
 }
 
+/**
+ * Returns the sum of the numbers in array.
+ *
+ * @param array The array of number to iterate over
+ */
+export function sum(array: number[]): number {
```
Comment (Rhombus33) [inline src/utils/misc.ts:28]: test for these helper methods?

File: src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+      const text = `The average active users of the <b>${ordinalTexts[i + 1]} week</b> was <b>${percentageChangeText}</b> the <b>${ordinalTexts[i]} week</b>.`;
+
+      summaries.push({
+        text,
+        validity: 1.0,
+      });
+    }
+
+    return summaries;
```
Comment (Rhombus33) [inline src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts:39]: Tests?

File: src/utils/misc.ts (hunk 14 above)
Comment (jcc37) [reply src/utils/misc.ts:28]: This helper method will be removed and replaced with `mathjs.sum`

File: src/datasets/summarizations/libs/trend.ts (hunk 6 above)
Comment (jcc37) [reply src/datasets/summarizations/libs/trend.ts:178]: Will rename to additiveSeasonalDecompose

File: src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts (hunk 13 above)
Comment (jcc37) [reply src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts:17]: I prefer to keep the prefix `week` to align with other variables like `weekPoints, but I just renamed it to `weekYAverages` to improve readability.

File: src/datasets/summarizations/libs/trend.ts (hunk 5 above)
Comment (jcc37) [reply src/datasets/summarizations/libs/trend.ts]: I added JSDoc comments for `additiveDecomposite` and explained the relationships between the inputs and outputs there.

File: src/datasets/summarizations/libs/trend.ts (hunk 4 above)
Comment (jcc37) [reply src/datasets/summarizations/libs/trend.ts:240]: The explanation of parameter `groupFn`, which is the only place that uses the type `GroupIdentifier`, has been added in the JSDoc comment of `additiveDecomposite`.

File: src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts (hunk 11 above)
Comment (jcc37) [reply src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts:14]: Weeks with < 3 days omitted in this summary. Any suggestions on improving wording and the comments?

Comment (Rhombus33) [reply src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts:14]: "Weeks with 3 days or less are considered to belong to the last/next 30 days and are omitted from the summaries."

What defines a week here? Is it possible that you omit 2 partial weeks (one in the beginning and one at the end)? Or if it is just by number of days, do you always omit the beginning or end?

Comment (jcc37) [reply src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts:14]: > Is it possible that you omit 2 partial weeks (one in the beginning and one at the end)?
Yes.

Considering weeks with <= 3 days may make the average of y-values in a week noisy and further harm the quality of the summaries. For 28-day data (GA data), there are always 4 valid weeks based on this rule. But for 30-day or 31-day data, there might be 4 or 5 weeks to create summaries.

File: src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts (hunk 10 above)
Comment (jcc37) [reply src/datasets/summarizations/trend-weekly-comparison-average.summarization.ts:11]: Will copy the part of the description and sample summaries from the design doc in the refactor PR.

**Review by Rhombus33 — APPROVED** [review]

**Pull request merged**